playbooks/hail_batch.md

```diff
@@ -0,0 +1,108 @@
+# Hail: Batch playbook
+
+## Batch
+
+### Tool reports out of memory, but VM resources don't agree?
+
+Some tools run inside their own virtual machine, inside the virtual machine (like docker on your computer). You can configure these tools to:
+
+- Use the system memory,
+    - but often they don't know how much system memory there is, sometimes they keep allocating if they need it, without releasing memory back.
+- Use a specific [lower, upper] amount of memory:
+    - Sometimes this gets pre-allocated, for example Java with the `-xms` flag will immediately appear to consume that memory, because it's reserving it for the task.
+    - Sometimes it uses it will allocate as required, up to some known limit (`-xmx`). This means Java may report out of memory, even though the host machine still has more memory to consume.
+
+Linked discussions:
+
+- [https://centrepopgen.slack.com/archives/C030X7WGFCL/p1712182771574679](https://centrepopgen.slack.com/archives/C030X7WGFCL/p1712182771574679)
+
+
+
+## Python jobs
+
+
+Context:
+
+Hail Batch Python jobs are a little bit of black magic. It uses the python packages dill and pickle to capture all the scope that a function needs. This could include:
+
+  - `inputs`: this is done by pickling the value in a very generic way, so is usually very space and memory inefficient,
+  - `outputs`: similar to inputs,
+  - `imports` (especially for type annotations)
+      - This can be very bad, as some imports (like subprocess) are very system dependent, and trying to pickle and reattach an import on a new system will almost certainly fail,
+  - _variables declared outside a function_
+  - any other runtime information it needs,
+
+This makes for a really smooth user experience when it works, but a nightmare to debug.
+
+Pro tips, make sure your function that you're calling is as self sufficient as possible, consider:
+
+  - REMOVING type annotations
+  - ensuring all imports are INSIDE the function
+  - NOT passing fully formed objects (dataframes, matrix tables, lists [which don't work anyway]),
+      - instead write the file to cloud storage in an efficient format,
+      - pass the PATH to the python job, and reload.
+
+### Runs out of memory on job start-up
```

Comment:
may be worth flagging that this can also manifest as a driver job to set up a batch running for hours, even if it does succeed

Reply:
```suggestion
### Runs out of memory on job start-up / set-up job takes a very long time
```